regression in caf::io::spawn_client

While we didn't really touch the I/O module recently, something broke `spawn_client`. Code to reproduce:

```
  sys.middleman().spawn_client([](caf::io::broker*) {}, "foo", 42);
```

### Discussed in https://github.com/actor-framework/actor-framework/discussions/1899

<div type='discussions-op-text'>

<sup>Originally posted by **xliao-neology** June 27, 2024</sup>
we been using spawn_client() perfectly in 0.19.6. However same code will have below error show up in 1.0.0.
I understand the trend is to use caf_net. But for now, any way that I can do to keep spawn_client() work with 1.0.0 release?

>>>>>
/usr/local/include/caf/actor_storage.hpp:30:5: error: ‘caf::io::abstract_broker::abstract_broker(caf::actor_config&)’ is protected within this context
   30 |     new (&data) T(std::forward<Us>(zs)...);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
</div>

--- a/libcaf_io/CMakeLists.txt
+++ b/libcaf_io/CMakeLists.txt
@@ -36,6 +36,7 @@ caf_add_component(
     caf/io/datagram_servant.cpp
     caf/io/doorman.cpp
     caf/io/middleman.cpp
+    caf/io/middleman.test.cpp
     caf/io/middleman_actor.cpp
     caf/io/middleman_actor_impl.cpp
     caf/io/system_messages.test.cpp

--- a/libcaf_io/caf/io/abstract_broker.hpp
+++ b/libcaf_io/caf/io/abstract_broker.hpp
@@ -74,6 +74,8 @@ public:
   ~abstract_broker() override;
 
   // even brokers need friends
+  template <class T>
+  friend class caf::actor_storage;
   friend class scribe;
   friend class doorman;
   friend class datagram_servant;

--- a/libcaf_io/caf/io/middleman.test.cpp
+++ b/libcaf_io/caf/io/middleman.test.cpp
@@ -12,15 +12,18 @@ using namespace caf::io;
 
 namespace {
 
-// Regression test for GH-#1900.
-// Note: we can't reproduce this in compile time,
-// static_assert(std::is_same_v<..snippet.., expected<actor>>) won't work.
-TEST("spawn client works") {
-  caf::actor_system_config cfg;
-  cfg.load<io::middleman>();
-  caf::actor_system sys{cfg};
-  sys.middleman().spawn_client(
-    [](caf::io::broker*, caf::io::connection_handle) {}, "foo", 42);
+TEST("GH-1900 regression") {
+  // Note: we don't need to actually run `spawn_client`, we only check for a
+  // compiler error. We shouldn't actually run this code, since it will assume
+  // 42 is a socket descriptor.
+  auto fn = [] {
+    caf::actor_system_config cfg;
+    cfg.load<io::middleman>();
+    caf::actor_system sys{cfg};
+    sys.middleman().spawn_client(
+      [](caf::io::broker*, caf::io::connection_handle) {}, "foo", 42);
+  };
+  check(std::is_same_v<decltype(fn), decltype(fn)>);
 }
 
 } // namespace

--- a/libcaf_io/caf/io/middleman.test.cpp
+++ b/libcaf_io/caf/io/middleman.test.cpp
@@ -1,0 +1,26 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/main/LICENSE.
+
+#include "caf/io/middleman.hpp"
+
+#include "caf/test/fixture/deterministic.hpp"
+#include "caf/test/test.hpp"
+
+using namespace caf;
+using namespace caf::io;
+
+namespace {
+
+// Regression test for GH-#1900.
+// Note: we can't reproduce this in compile time,
+// static_assert(std::is_same_v<..snippet.., expected<actor>>) won't work.
+TEST("spawn client works") {
+  caf::actor_system_config cfg;
+  cfg.load<io::middleman>();
+  caf::actor_system sys{cfg};
+  sys.middleman().spawn_client(
+    [](caf::io::broker*, caf::io::connection_handle) {}, "foo", 42);
+}
+
+} // namespace

--- a/libcaf_io/main.test.cpp
+++ b/libcaf_io/main.test.cpp
@@ -4,4 +4,6 @@
 
 #include "caf/test/caf_test_main.hpp"
 
-CAF_TEST_MAIN()
+#include "caf/io/middleman.hpp"
+
+CAF_TEST_MAIN(caf::io::middleman)

